Pull HySDS Docker Images from a Remote Registry + Refactor HySDS Cluster Terraform Module into a Sub-module

drewm-jpl opened the pull request:
- Pull HySDS Docker images from GHCR/any container registry. Uses `kubernetes_secret` Terraform resource to authenticate with GHCR. 
  - This requires creating a personal access token in Github. Once created,  you must pass it as the `container_registry_password` variable within `terraform.tfvars` along with your Github username as the `container_registry_username` variable.
- Modularize the HySDS Terraform cluster to be able to be invoked as a submodule from a root module.
- Invoke the HySDS cluster Terraform module from a remote source.
  - Currently it is sourced from the GitHub repo. Eventually, we will probably want to source it from another location or a package release on Github, Artifactory, etc.
  - To be able to pull a Terraform module from a Github repository source, the branch you're pulling from has to have a different name than any file or directory in the base of your repository. For this reason, I renamed both the branch and the Terraform directory within the repo. Previously, they were both named `terraform`.
- Specify Terraform version to be used within the repo in `.terraform-version`. This change was made to enable `tfenv` to automatically use the repo-specified Terraform version. This enables developers to ensure they are using the 
  correct version of Terraform.
- _Note:_ The commit history may be a bit funky because I accidentally checked in my personal access token at one point and needed to scrub that commit from the history (whoops). I've also since deleted that token.

resolves #55

LucaCinquini commented:
Hi Drew, I approved the PR but could you please make the following changes:

o Update the README file for starting the cluster stating that:
- the use needs to cd to the terraform-unity-sps-root-module directory first
- the user needs to create a "terraform.tfvars" file, and can you show an example content (without sensitive information)

o I think the main.tf file needs to be updated to reference the correct directory for the module, and branch (which after the merge should be "main")

o It seems like "terraform apply" will crash if the kubernetes namespace already exists. Is there anyway to avoid that (if, for example, other resources are already deployed in that namespace)?

Thanks again,
Luca

drewm-jpl commented:
> Hi Drew, I approved the PR but could you please make the following changes:
> 
> 1. Update the README file for starting the cluster stating that:
> 
> * the use needs to cd to the terraform-unity-sps-root-module directory first
> * the user needs to create a "terraform.tfvars" file, and can you show an example content (without sensitive information)
> 
> 2. I think the main.tf file needs to be updated to reference the correct directory for the module, and branch (which after the merge should be "main")
> 
> 3. It seems like "terraform apply" will crash if the kubernetes namespace already exists. Is there anyway to avoid that (if, for example, other resources are already deployed in that namespace)?
> 
> Thanks again, Luca

Hi Luca,

Thanks for your feedback. Here's my response to the points you raised:

- I've addressed points 1 and 2 with [my most recent commit](https://github.com/unity-sds/unity-sps-prototype/pull/78/commits/b66c09aa9f5c649488b887b74dad69ee0a44efeb).

- To address point 3, I think we should first discuss how modular we would like our SPS deployment to be. For example, if the the HySDS cluster and WPST API are intended to be deployed separately, then perhaps we should reconsider creating the namespace at the submodule level. Instead, we could create it at the root module level and then have the submodules inherit the namespace from a [Terraform `data source`](https://www.terraform.io/language/data-sources). This [stackoverflow post](https://stackoverflow.com/questions/59491608/in-terraform-0-12-how-to-skip-creation-of-resource-if-resource-name-already-ex) has a relevant discussion on this topic.

Regards,
Drew
